### PR TITLE
button changes

### PR DESCRIPTION
**Pull request by SA-22C**
add event EVENT_HELD_MEDIUM (used in my prop file)
change checks in button_base.h to limit false activation of HELD state when rapidly pressing a button.  Helps when trying to spam blaster blocks.

### CODE REVIEW

**Comment by SA-22C** [comment]
Other changes include setting a time duration for long press vs short press so we can differentiate them a little better.

File: props/prop_base.h [hunk 1]
```diff
@@ -207,24 +207,24 @@ class PropBase : CommandParser, Looper, protected SaberBase {
       current_preset_.variation = SaberBase::GetCurrentVariation();
       current_preset_.Save();
     }
-#endif	
+#endif
```
Comment (profezzorn) [inline props/prop_base.h]: None of the changes in this file seem relevant to the button changes.

File: buttons/button_base.h [hunk 2]
```diff
@@ -30,27 +31,36 @@ class ButtonBase : public Looper,
...
-      }
-      while (DebouncedRead()) YIELD();
-      prop.Event(button_, EVENT_RELEASED);
-      if (current_modifiers & button_) {
-        current_modifiers &=~ button_;
-        if (millis() - push_millis_ < 500) {
-          prop.Event(button_, EVENT_CLICK_SHORT);
-        } else {
-          prop.Event(button_, EVENT_CLICK_LONG);
+          if (!short_click){
```
Comment (profezzorn) [inline buttons/button_base.h]: short click will ALWAYS be false here

File: buttons/button_base.h [hunk 3]
```diff
@@ -78,7 +88,12 @@ class ButtonBase : public Looper,
   const char* name_;
   enum BUTTON button_;
   uint32_t push_millis_;
+  bool short_click;
```
Comment (profezzorn) [inline buttons/button_base.h]: global variables should end in _

File: buttons/button_base.h [hunk 4]
```diff
@@ -30,27 +31,36 @@ class ButtonBase : public Looper,
...
-      prop.Event(button_, EVENT_RELEASED);
-      if (current_modifiers & button_) {
-        current_modifiers &=~ button_;
-        if (millis() - push_millis_ < 500) {
-          prop.Event(button_, EVENT_CLICK_SHORT);
-        } else {
-          prop.Event(button_, EVENT_CLICK_LONG);
+          if (!short_click){
+            if (millis() - push_millis_ > 300) {
+                prop.Event(button_, EVENT_HELD);
```
Comment (profezzorn) [inline buttons/button_base.h]: indentation should be two spaces per level, not four

File: props/prop_base.h (hunk 1 above)
Comment (SA-22C) [reply props/prop_base.h]: I think my editor did a bunch of whitespace garbage.  I'll try to correct it.

File: buttons/button_base.h (hunk 2 above)
Comment (SA-22C) [reply buttons/button_base.h]: No, short_click is set to true when a subsequent short click occurs within 500 ms of the first.  this way if we are a bit long on the press but release and press the button again (like one does when spamming blaster blocks) the event_held doesn't fire and we dont' get false held events.  I just did a test with line 60 commented out vs not: when short_click is set to true, the false event_held events don't show up on the console.  When that line is commented and short_click is always false, event_held events are shown in the log.

File: buttons/button_base.h [hunk 5]
```diff
@@ -30,27 +31,36 @@ class ButtonBase : public Looper,
         current_modifiers |= button_;
       }
       while (DebouncedRead()) {
```
Comment (profezzorn) [inline buttons/button_base.h]: This loop seems to have been extended further than it used to be. Is that intentional?

File: buttons/button_base.h [hunk 6]
```diff
@@ -30,31 +30,37 @@ class ButtonBase : public Looper,
...
-	}
-	YIELD();
+        if (millis() - push_millis_ > 300) {
+            prop.Event(button_, EVENT_HELD);
+          while (DebouncedRead()) {
+            if (millis() - push_millis_ > 800){
+              prop.Event(button_, EVENT_HELD_MEDIUM);
+              while (DebouncedRead()) {
+                if (millis() - push_millis_ > 2000) {
+                    prop.Event(button_, EVENT_HELD_LONG);
```
Comment (profezzorn) [inline buttons/button_base.h]: should be indented 2 steps

File: buttons/button_base.h [hunk 7]
```diff
@@ -30,31 +30,37 @@ class ButtonBase : public Looper,
...
-	    if (millis() - push_millis_ > 2000) {
-	      prop.Event(button_, EVENT_HELD_LONG);
-	      while (DebouncedRead()) YIELD();
-	    }
-	    YIELD();
-	  }
-	}
-	YIELD();
+        if (millis() - push_millis_ > 300) {
+            prop.Event(button_, EVENT_HELD);
```
Comment (profezzorn) [inline buttons/button_base.h]: should be indented 2 steps

File: props/prop_base.h (hunk 1 above)
Comment (SA-22C) [reply props/prop_base.h]: fixed whitespace on prop_base.h

File: buttons/button_base.h [hunk 8]
```diff
@@ -30,31 +30,37 @@ class ButtonBase : public Looper,
...
-      if (current_modifiers & button_) {
-        current_modifiers &=~ button_;
-        if (millis() - push_millis_ < 500) {
-          prop.Event(button_, EVENT_CLICK_SHORT);
+        prop.Event(button_, EVENT_RELEASED);
+        if (current_modifiers & button_) {
+            current_modifiers &=~ button_;
+          if (millis() - push_millis_ < 500) {
+            prop.Event(button_, EVENT_CLICK_SHORT);
+          } else if (millis() - push_millis_ > 500 && millis() - push_millis_ < 2500){
```
Comment (profezzorn) [inline buttons/button_base.h]: space between ) and {

File: buttons/button_base.h [hunk 9]
```diff
@@ -30,31 +30,37 @@ class ButtonBase : public Looper,
...
+        }
+        YIELD();
       }
       while (DebouncedRead()) YIELD();
-      prop.Event(button_, EVENT_RELEASED);
-      if (current_modifiers & button_) {
-        current_modifiers &=~ button_;
-        if (millis() - push_millis_ < 500) {
-          prop.Event(button_, EVENT_CLICK_SHORT);
+        prop.Event(button_, EVENT_RELEASED);
```
Comment (profezzorn) [inline buttons/button_base.h]: extra indentation

File: buttons/button_base.h (hunk 2 above)
Comment (SA-22C) [reply buttons/button_base.h]: well that's embarrassing.  Could have sworn it made a difference in testing but just checked again with some prints and it never flags true.  Removing.

File: buttons/button_base.h (hunk 3 above)
Comment (SA-22C) [reply buttons/button_base.h]: removed as it doesn't perform the check I thought it did.

File: buttons/button_base.h (hunk 4 above)
Comment (SA-22C) [reply buttons/button_base.h]: fixed.

File: buttons/button_base.h (hunk 5 above)
Comment (SA-22C) [reply buttons/button_base.h]: Not sure what you mean.

File: buttons/button_base.h (hunk 7 above)
Comment (SA-22C) [reply buttons/button_base.h]: fixed.

File: buttons/button_base.h (hunk 6 above)
Comment (SA-22C) [reply buttons/button_base.h]: fixed

File: buttons/button_base.h (hunk 8 above)
Comment (SA-22C) [reply buttons/button_base.h]: fixed

File: buttons/button_base.h (hunk 9 above)
Comment (SA-22C) [reply buttons/button_base.h]: fixed.

File: buttons/button_base.h [hunk 10]
```diff
@@ -30,28 +30,34 @@ class ButtonBase : public Looper,
...
+        }
+        YIELD();
       }
       while (DebouncedRead()) YIELD();
       prop.Event(button_, EVENT_RELEASED);
-      if (current_modifiers & button_) {
-        current_modifiers &=~ button_;
-        if (millis() - push_millis_ < 500) {
-          prop.Event(button_, EVENT_CLICK_SHORT);
+        if (current_modifiers & button_) {
```
Comment (profezzorn) [inline buttons/button_base.h]: extra indentation (here and below)

File: buttons/button_base.h [hunk 11]
```diff
@@ -30,28 +30,34 @@ class ButtonBase : public Looper,
...
-	    if (millis() - push_millis_ > 2000) {
-	      prop.Event(button_, EVENT_HELD_LONG);
-	      while (DebouncedRead()) YIELD();
-	    }
-	    YIELD();
-	  }
-	}
-	YIELD();
+        if (millis() - push_millis_ > 300) {
+            prop.Event(button_, EVENT_HELD);
```
Comment (profezzorn) [inline buttons/button_base.h]: why is this line indented more than the next one?

**Pull request merged**